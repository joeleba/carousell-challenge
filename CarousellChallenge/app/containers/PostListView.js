import React, { Component } from 'react'
import { AppRegistry, Text, FlatList, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux'
import { NavigationActions } from 'react-navigation';
import { addPost } from '../actions'
import NewContentForm from '../components/NewContentForm'

class PostListView extends Component {
  _keyExtractor(item, index) {
    return item.title;
  }

  render() {
    console.log(this.props.posts);
    return (
      <View style={styles.container}>
      <Text>HELOOOOOO</Text>
        <FlatList
          data={this.props.posts}
          keyExtractor={this._keyExtractor}
          renderItem={({item}) =>
          <Text
            style={styles.item}
            onPress={() => this.props.dispatchNavigateToPost(item.id)}>
            {item.title}
          </Text>
        }
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
   flex: 1,
   paddingTop: 22
  },
  item: {
    padding: 10,
    fontSize: 18,
    height: 44,
  },
})

function mapStateToProps(state) {
  return {
    posts: state.contentReducers.posts
  }
}

function mapDispatchToProps(dispatch) {
  return {
    dispatchNavigateToPost: (postId) =>
      dispatch(NavigationActions.navigate({
        routeName: 'PostView',
        params: {
          postId: postId
        }
      }))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(PostListView)
