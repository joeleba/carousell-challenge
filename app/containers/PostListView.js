import React, { Component } from 'react';
import { Text, FlatList, ScrollView, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import Icon from '@expo/vector-icons/FontAwesome';
import _ from 'lodash';

import { upvote, downvote, POST_TYPE } from '../actions';

class PostListView extends Component {
  keyExtractor(item, index) {
    return item.id;
  }

  render() {
    // I'm aware that the problem asked us to sort the topic by "upvotes"
    // I assume that you meant the "overral vote count" i.e. upvote - downvote
    let posts = _.orderBy(
      _.values(this.props.postsObject),
      [(p) => { return p.upvoteCount - p.downvoteCount; }],
      ['desc']
    );

    return (
      <View style={styles.container}>
        <ScrollView>
          <FlatList
            data={posts}
            keyExtractor={this.keyExtractor}
            renderItem={({ item }) =>
              (<View style={styles.item}>
                <Text style={styles.vote}>
                  <Icon
                    name="arrow-up"
                    onPress={() => this.props.dispatchUpvoteAction(item.id)}
                  />
                  {`  ${(item.upvoteCount - item.downvoteCount)}  `}
                  <Icon
                    name="arrow-down"
                    onPress={() => this.props.dispatchDownvoteAction(item.id)}
                  />
                </Text>
                <Text
                  style={styles.title}
                  onPress={() => this.props.dispatchNavigateToPost(item.id)}
                >
                  {item.title}
                </Text>
              </View>)
            }
          />
        </ScrollView>

        <View style={styles.floatingNewPostButton}>
          <Icon
            name="plus"
            backgroundColor="#3b5998"
            size={20}
            onPress={() => this.props.dispatchNavigateToNewPost()}
          />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: 22,
  },
  title: {
    fontSize: 18,
    flex: 0.8,
  },
  floatingNewPostButton: {
    alignSelf: 'flex-end',
    position: 'absolute',
    bottom: 35,
    right: 35,
  },
  vote: {
    fontSize: 18,
    flex: 0.2,
    color: 'gray',
  },
  item: {
    borderWidth: 1,
    borderColor: 'lightgrey',
    flexDirection: 'row',
    paddingVertical: 20,
    paddingHorizontal: 10,
  },
});

function mapStateToProps(state) {
  return {
    postsObject: state.contentReducers.posts,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    dispatchNavigateToPost: postId =>
      dispatch(NavigationActions.navigate({
        routeName: 'PostView',
        params: {
          postId,
        },
      })),
    dispatchNavigateToNewPost: () =>
      dispatch(NavigationActions.navigate({ routeName: 'NewPost' })),
    dispatchUpvoteAction: id => dispatch(upvote(id, POST_TYPE)),
    dispatchDownvoteAction: id => dispatch(downvote(id, POST_TYPE)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(PostListView);
