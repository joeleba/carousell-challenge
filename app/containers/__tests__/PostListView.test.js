import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import PostListView from '../PostListView';

const mockStore = configureStore();

const stateOnePost = {
  contentReducers: {
    currentId: 1,
    posts: {
      0: {
        children: [],
        content: "ccc",
        downvoteCount: 0,
        upvoteCount: 0,
        title: "sad",
        id: 0
      }
    },
    comments: {}
  }
};

const stateNoPost = {
  contentReducers: {
    currentId: 0,
    posts: {},
    comments: {}
  }
};

describe('Component: PostListView', () => {
  it('renders with correct postsObject prop', () => {
    const wrapper = shallow(
      <PostListView />,
      { context: { store: mockStore(stateOnePost) } },
    );
    expect(wrapper.dive()).toMatchSnapshot();
    expect(wrapper.prop('postsObject')).toEqual({
      0: {
        children: [],
        content: "ccc",
        downvoteCount: 0,
        upvoteCount: 0,
        title: "sad",
        id: 0
      }
    });
  });

  it('renders when there\'s no post', () => {
    const wrapper = shallow(
      <PostListView />,
      { context: { store: mockStore(stateNoPost) } },
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });
});
