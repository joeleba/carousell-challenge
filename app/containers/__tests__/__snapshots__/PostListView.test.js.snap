// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: PostListView renders when there's no post 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 22,
    }
  }
>
  <ScrollView>
    <FlatList
      data={Array []}
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </ScrollView>
  <View
    style={
      Object {
        "alignSelf": "flex-end",
        "bottom": 35,
        "position": "absolute",
        "right": 35,
      }
    }
  >
    <Icon
      allowFontScaling={false}
      backgroundColor="#3b5998"
      name="plus"
      onPress={[Function]}
      size={20}
    />
  </View>
</View>
`;

exports[`Component: PostListView renders with correct postsObject prop 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingTop": 22,
    }
  }
>
  <ScrollView>
    <FlatList
      data={
        Array [
          Object {
            "children": Array [],
            "content": "ccc",
            "downvoteCount": 0,
            "id": 0,
            "title": "sad",
            "upvoteCount": 0,
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </ScrollView>
  <View
    style={
      Object {
        "alignSelf": "flex-end",
        "bottom": 35,
        "position": "absolute",
        "right": 35,
      }
    }
  >
    <Icon
      allowFontScaling={false}
      backgroundColor="#3b5998"
      name="plus"
      onPress={[Function]}
      size={20}
    />
  </View>
</View>
`;
